activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -159,6 +159,11 @@ def begin_transaction(options = {})
 
       def commit_transaction
         inner_transaction = @stack.pop
+
+        unless current_transaction.joinable?
+          inner_transaction.records.each(&:call_before_commit)
+        end
+
```

Comment:
The txn manager is reaching into the txn and doing internal work. It knows that the txn has records and that the records have before-commit callbacks — much more internal knowledge than it needs. Consider telling the txn only what we know, that we're about to commit records at a transaction boundary (e.g. `inner_transaction.before_commit_records`)